rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,298 @@
+    // ClusterConfig.SessionRecording is renamed to SessionRecordingConfig.Mode.
+    string Mode;
+    bool ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// DisconnectExpiredCert & LocalAuth fields are moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool AllowLocalAuth;
+    bool DisconnectExpiredCert;
+}
+```
+
+#### Data migration & docs updates
+
+A new step should be added to `migrateLegacyResources` (called during auth
+server initialization) that takes care of migrating the `ClusterConfig` data
+stored in the backend to the new resource structures.  This probably implies
+that `ClusterConfig` as a structure needs to be kept around for a couple of
+subsequent Teleport versions for marshalling purposes.
+
+Documentation should be updated accordingly.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `mode` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
+1. The value of `mode` can be changed only if its original value
+   is one of `off`, `node` or `node-sync`.
+1. The value of `mode` can be changed to only one of `off`, `node`
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no_proxy_recording`. A user could then
+be assigned a role featuring `update.no_proxy_recording` instead of the regular, more
+permissive `update`:
+```
+  allow:
+    rules:
+    - resources: [session_recording_config]
+      verbs: [read, update.no_proxy_recording]
+```
+
+### Additional dynamically configurable resources
+
+In addition to the resources derived from `ClusterConfig`, the resources
+`ClusterAuthPreference` and `PAMConfig` should also be adapted to facilitate
+the dynamic configuration workflow.
+
+### Working with a whole cluster configuration
+
+`KindClusterConfig` should be retained but reinterpreted as a helper meta-kind
+similar to `KindConnectors`. It would allow aggregating all the cluster config
+related resources into a single `tctl get` output:
+
+```
+$ tctl get cluster_config
+kind: session_recording_config
+[...]
+---
+kind: cluster_networking_config
+[...]
+---
+kind: audit_config
+[...]
+---
+kind: cluster_auth_preference
```

Comment:
is that already a new resource? the name `cluster_auth_config` will work better as it will match every other one